examples/dynamic_api.rs

```diff
@@ -19,59 +19,41 @@ async fn main() -> anyhow::Result<()> {
 
     let ns_filter = std::env::var("NAMESPACE").ok();
 
-    // The following loops turn the /api or /apis listers into kube::api::Resource
-    // objects which can be used to make dynamic api calls.
-    // This is slightly awkward because of corev1 types
-    // and data split over the list types and the inner get calls.
-
-    // loop over all api groups (except core v1)
-    let apigroups = client.list_api_groups().await?;
-    for g in apigroups.groups {
-        warn!("api group: {}", g.name);
-        let ver = g
-            .preferred_version
-            .as_ref()
-            .or_else(|| g.versions.first())
-            .expect("preferred or versions exists");
-        let apis = client.list_api_group_resources(&ver.group_version).await?;
-        print_group(&client, &ver.group_version, apis, ns_filter.as_deref()).await?;
-    }
-
-    warn!("core/v1 legacy group");
-    let coreapis = client.list_core_api_versions().await?;
-    assert_eq!(coreapis.versions.len(), 1);
-    let corev1 = client.list_core_api_resources(&coreapis.versions[0]).await?;
-    print_group(&client, &coreapis.versions[0], corev1, ns_filter.as_deref()).await
-}
-
-async fn print_group(
-    client: &Client,
-    group_version: &str,
-    apis: APIResourceList,
-    ns_filter: Option<&str>,
-) -> anyhow::Result<()> {
-    for ar in apis.resources {
-        if !ar.verbs.contains(&"list".to_string()) {
-            continue;
-        }
-        let gvk = GroupVersionKind::from_api_resource(&ar, &apis.group_version);
-        let api: Api<DynamicObject> = if ar.namespaced {
-            if let Some(ns) = ns_filter {
-                Api::namespaced_with(client.clone(), ns, &gvk)
+    let discovery = Discovery::new(&client).await?;
+
+    for group in discovery.groups() {
+        let ver = group.preferred_version();
+        for gvk in group.resources_by_version(ver) {
+            let kind = DynamicObject::kind(&gvk);
+            let (_, raw_resource) = discovery
+                .resolve_group_version_kind(group.name(), ver, &kind)
+                .unwrap();
+            let api: Api<DynamicObject> = if raw_resource.namespaced {
+                if let Some(ns) = &ns_filter {
+                    Api::namespaced_with(client.clone(), ns, &gvk)
+                } else {
+                    Api::all_with(client.clone(), &gvk)
+                }
             } else {
                 Api::all_with(client.clone(), &gvk)
+            };
+
+            info!("{}/{} : {}", group.name(), ver, kind);
+
+            let list = match api.list(&Default::default()).await {
+                Ok(l) => l,
+                Err(e) => {
+                    warn!("Failed to list: {:#}", e);
+                    continue;
```

Comment:
is there a reason this is expected to fail?

Reply:
There are non-listable API resources, such as authorization-related ones:
```json
# kubectl get --raw /apis/authorization.k8s.io/v1 | jq -r
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "authorization.k8s.io/v1",
  "resources": [
    {
      "name": "localsubjectaccessreviews",
      "singularName": "",
      "namespaced": true,
      "kind": "LocalSubjectAccessReview",
      "verbs": [
        "create"
      ]
    },
    {
      "name": "selfsubjectaccessreviews",
      "singularName": "",
      "namespaced": false,
      "kind": "SelfSubjectAccessReview",
      "verbs": [
        "create"
      ]
    },
    {
      "name": "selfsubjectrulesreviews",
      "singularName": "",
      "namespaced": false,
      "kind": "SelfSubjectRulesReview",
      "verbs": [
        "create"
      ]
    },
    {
      "name": "subjectaccessreviews",
      "singularName": "",
      "namespaced": false,
      "kind": "SubjectAccessReview",
      "verbs": [
        "create"
      ]
    }
  ]
}
```
(You can see that `list` verb is missing)